### PR TITLE
Use allocation ids to prevent repeated recovery of failed shards

### PR DESCRIPTION
**Pull request by ywelsch**
code simplification

### CODE REVIEW

**Comment by s1monw** [comment]
WOW awesome. LGTM but I think @bleskes will need to look too

**Comment by ywelsch** [comment]
I simplified the loop a bit (so we do not need goto to outer loop). can you have a look @bleskes?

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 1]
```diff
@@ -510,33 +489,28 @@ private void cleanFailedShards(final ClusterChangedEvent event) {
...
+
         while (iterator.hasNext()) {
-            Map.Entry<ShardId, FailedShard> entry = iterator.next();
-            FailedShard failedShard = entry.getValue();
-            IndexRoutingTable indexRoutingTable = routingTable.index(entry.getKey().getIndex());
+            Map.Entry<ShardId, ShardRouting> entry = iterator.next();
+            ShardId failedShardId = entry.getKey();
+            ShardRouting failedShardRouting = entry.getValue();
+            IndexRoutingTable indexRoutingTable = routingTable.index(failedShardId.getIndex());
             if (indexRoutingTable != null) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: we should remove the iterator in this case. I would just do:

```
if (indexRoutingTable == null) {
  iterator.remove();
  continue;
}
```

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 2]
```diff
@@ -510,33 +489,28 @@ private void cleanFailedShards(final ClusterChangedEvent event) {
...
-                IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(entry.getKey().id());
+                IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(failedShardId.id());
                 if (shardRoutingTable != null) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: same here re iterator.remove()

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 3]
```diff
@@ -510,33 +489,28 @@ private void cleanFailedShards(final ClusterChangedEvent event) {
...
                     for (ShardRouting shardRouting : shardRoutingTable.assignedShards()) {
                         if (localNodeId.equals(shardRouting.currentNodeId())) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: I don't think we need the filtering on the node. How about using:

```
shardRoutingTable.assignedShards().stream().filter(as -> as.isSameAllocation(failedShard)).findFirst().isPresent()
```

If true, we keep the shard failure, o.w. we clear the iterator...

**Comment by bleskes** [comment]
This is great! I left some comments simplifying it even further..

**Comment by ywelsch** [comment]
Thanks @bleskes. Java 8 streams to the rescue. Looks way better! I made one minor improvement to your suggestion:
```
if (shardRoutingTable.assignedShards().stream()
        .noneMatch(shr -> shr.isSameAllocation(failedShardRouting))) {
    iterator.remove();
}
```

**Comment by bleskes** [comment]
That noneMatch is sexy indeed. LGTM. thanks @ywelsch .

**Pull request closed without merging**